accord-core/src/test/java/accord/impl/basic/SimulatedDelayedExecutorService.java

```diff
@@ -26,12 +31,105 @@
 import accord.burn.random.RandomWalkRange;
 import accord.utils.RandomSource;
 
-public class SimulatedDelayedExecutorService extends TaskExecutorService
+import static java.util.concurrent.TimeUnit.NANOSECONDS;
+
+public class SimulatedDelayedExecutorService extends TaskExecutorService implements ScheduledExecutorService
 {
+    private class ScheduledTask<T> extends Task<T> implements ScheduledFuture<T>
+    {
+        private final long sequenceNumber;
+        private final long periodMillis;
+        private long timeMillis;
+        private ScheduledTask(long sequenceNumber, long value, TimeUnit unit, Callable<T> fn)
+        {
+            super(fn);
+            this.sequenceNumber = sequenceNumber;
+            periodMillis = unit.toMillis(value);
+        }
+
+        @Override
+        public long getDelay(TimeUnit unit)
+        {
+            return unit.convert(timeMillis - pending.nowInMillis(), TimeUnit.MILLISECONDS);
+        }
+
+        @Override
+        public int compareTo(Delayed other)
+        {
+            if (other == this) // compare zero if same object
+                return 0;
+            if (other instanceof ScheduledTask) {
+                ScheduledTask<?> x = (ScheduledTask<?>)other;
+                long diff = timeMillis - x.timeMillis;
+                if (diff < 0)
+                    return -1;
+                else if (diff > 0)
+                    return 1;
+                else if (sequenceNumber < x.sequenceNumber)
+                    return -1;
+                else
+                    return 1;
+            }
+            long diff = getDelay(NANOSECONDS) - other.getDelay(NANOSECONDS);
```

Comment:
I kind of prefer the appoach take in `java.util.concurrent` which is to immediately convert everything to `nanoseconds` to minimize the number of conversions and then only convert out of nanoseconds when displaying or making a call to another library.

Since everything also usually includes a time unit it just stays in `nanoseconds` indefinitely.

Why should this one diff here be in nanoseconds when the source unit is very clearly milliseconds? Seems like everything is milliseconds here.

Reply:
I honestly pulled this from `java.util.concurrent.ScheduledThreadPoolExecutor.ScheduledFutureTask#compareTo` and just changed it to match this type =D

This case is also the case where we don't know what the other `Delayed` type is, so the only thing we can actually do is ask it what the delay is, which requires a `TimeUnit`...

Here is the def of `Delayed` in java

```
public interface Delayed extends Comparable<Delayed> {

    /**
     * Returns the remaining delay associated with this object, in the
     * given time unit.
     *
     * @param unit the time unit
     * @return the remaining delay; zero or negative values indicate
     * that the delay has already elapsed
     */
    long getDelay(TimeUnit unit);
}
```

> when the source unit is very clearly milliseconds?

`this` is `milliseconds` but `other` is unknown... its possible that `other` is in `nanosecond` scale, so there would be a real `diff` if we convert `this` to `nanoseconds`...

Reply:
Ah my mistake. I didn't realize `Delayed` required was one of the standard ones requiring you to provide a unit.